internal/util-collection/src/main/scala/sbt/internal/util/Settings.scala

```diff
@@ -24,28 +26,73 @@ sealed trait Settings[Scope] {
 }
 
 private final class Settings0[Scope](
-    val data: Map[Scope, AttributeMap],
+    keyDefs: Map[AttributeKey[_], Map[Scope, Any]],
     val delegates: Scope => Seq[Scope]
 ) extends Settings[Scope] {
 
-  def scopes: Set[Scope] = data.keySet
+  // TODO: get rid of direct accesses to data completely, so it can be removed from the API
+  lazy val data: Map[Scope, AttributeMap] = {
+    keyDefs.toSeq
+      .flatMap {
+        case (key: AttributeKey[Any], map) =>
+          map.map { case (scope, value: Any) => (scope, AttributeEntry[Any](key, value)) }
+      }
+      .groupBy(_._1)
+      .map {
+        case (scope, entries) => (scope, AttributeMap(entries.map(_._2)))
+      }
+      .toMap
+  }
+  lazy val scopes: Set[Scope] = {
+    // TODO: optimize
+    keyDefs.flatMap(_._2.keySet).toSet
+  }
+  def attributeKeys: Set[AttributeKey[_]] = keyDefs.keySet
+
   def keys(scope: Scope) = data(scope).keys.toSet
 
   def allKeys[T](f: (Scope, AttributeKey[_]) => T): Seq[T] =
-    data.flatMap { case (scope, map) => map.keys.map(k => f(scope, k)) }.toSeq
-
-  def get[T](scope: Scope, key: AttributeKey[T]): Option[T] =
-    delegates(scope).toStream.flatMap(sc => getDirect(sc, key)).headOption
+    keyDefs.toIterator.flatMap {
+      case (key: AttributeKey[Any], map) =>
+        map.keys.map(scope => f(scope, key))
+    }.toSeq
+
+  def get[T](scope: Scope, key: AttributeKey[T]): Option[T] = {
```

Comment:
Would it make sense to keep the original `get` code as comment here since it's a one-liner?